sources/RuleRunner.cpp

```diff
@@ -1329,6 +1330,96 @@ ActionHasTarget(int8 type)
 }
 
 
+static bool
+getDirectoryPath(BString& str, const entry_ref& ref)
+{
+	BEntry entry(&ref);
+	if (entry.InitCheck() != B_OK)
+		return false;
+
+	if (!entry.IsDirectory() && entry.GetParent(&entry) != B_OK)
+		return false;
+
+	BPath path(&entry);
+	if (path.InitCheck() != B_OK)
+		return false;
+
+	str = path.Path();
+	return true;
+}
+
+
+bool
+SetTextForType(BString& text, int8 type, const entry_ref& ref, bool isTest)
+{
+	if (isTest)
+		switch (type) {
+			case TEST_TYPE:
+			{
+				BNode node(&ref);
+				if (node.InitCheck() != B_OK)
+					return false;
+
+				BNodeInfo nodeInfo(&node);
+				if (nodeInfo.InitCheck() != B_OK)
+					return false;
+
+				char mimeType[B_MIME_TYPE_LENGTH];
+				if (nodeInfo.GetType(mimeType) != B_OK)
+					return false;
+
+				text = mimeType;
+				return true;
+			}
+			case TEST_NAME:
+				text = ref.name;
+				return true;
+			case TEST_SIZE:
+			{
+				BEntry entry(&ref);
+				if (entry.InitCheck() != B_OK || !entry.IsFile())
+					return false;
+
+				off_t size;
+				if (entry.GetSize(&size) != B_OK)
+					return false;
+
+				text = "";
+				text << size;
+				return true;
+			}
+			case TEST_LOCATION:
+				return getDirectoryPath(text, ref);
+			default:
+				return false;
+		}
+	else
+		switch (type) {
+			case ACTION_MOVE:
+			case ACTION_COPY:
+				return getDirectoryPath(text, ref);
+			case ACTION_RENAME:
+				text = ref.name;
+				return true;
+			case ACTION_ARCHIVE:
+			{
+				BEntry entry(&ref);
+				if (entry.InitCheck() != B_OK || !entry.IsFile())
+					return false;
+
+				BPath path(&ref);
+				if (path.InitCheck() != B_OK)
+					return false;
+
+				text = path.Path();
+				return true;
```

Comment:
> (I am not 100% certain, but Filer only handles one file or folder per action. If so, Zip-O-Matic's naming the archive as "Archive.zip" by default when there are multiple files/folders does not apply here.)

You're totally right. I forgot about that...

Still, something's not right yet:
- Create a folder "theFolder" on the Desktop
- Create a rule: "When Name starts with tex, Do Add to archive" and drag&drop "theFolder" in the text control.
- Create a new textfile "text1" and drag&drop it into the DropZone.

One would expect an archive created as /boot/home/Desktop/theFolder/text1.zip.
We get /boot/home/Desktop/theFolder.zip.

If we add a "/" to the rules location (/boot/home/Desktop/theFolder/), we get an archive /boot/home/Desktop/theFolder/.zip.